soa-derive-internal/src/vec.rs

```diff
@@ -71,97 +88,144 @@ pub fn derive(input: &Input) -> TokenStream {
             /// ::capacity()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.capacity),
             /// the capacity of all fields should be the same.
             pub fn capacity(&self) -> usize {
-                let capacity = self.#first_field.capacity();
-                #(debug_assert_eq!(self.#fields_names_1.capacity(), capacity);)*
-                capacity
+                let vec: Vec<usize> = vec![#(self.data.#fields_names_1.capacity()),*];
```

Comment:
This should be an array, not a vector.

~~As for this code, the capacity of the different buffers may actually be different because we have no guarantee that the allocated size is the exact capacity we give them: as far as I know things like size and alignment of the type may change how much is allocated. If the first field allocates more than the rest we could use memory out of bounds.~~

It turns out that it's actually guaranteed. If the allocator (or the OS) allocates more memory, it gets ignored. We could use your implementation :)